### PR TITLE
Add teardown factory

### PR DESCRIPTION
**Pull request by RazTamir**
- This will help tearing down resources that were created as part of the test

Signed-off-by: ratamir <ratamir@redhat.com>

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -269,12 +271,12 @@ def factory(
                 wait=False
             )
             assert pvc_obj, "Failed to create PVC"
+            instances.append(pvc_obj)
+            pvc_obj.storageclass = storageclass
```
Comment (fbalak) [inline tests/conftest.py]: I would keep this on its original place so it can create reference if object is provided in parameter. It provides consistency for calling the factory. Also if we want to remove this reference for custom_data it should be done in all fixtures.

File: tests/conftest.py [hunk 2]
```diff
@@ -101,7 +101,8 @@ def factory(interface=constants.CEPHBLOCKPOOL):
             ).get(defaults.CEPHFILESYSTEM_NAME)
             ceph_pool_obj = OCS(**cfs)
         assert ceph_pool_obj, f"Failed to create {interface} pool"
-        instances.append(ceph_pool_obj)
+        if interface == constants.CEPHBLOCKPOOL:
```
Comment (fbalak) [inline tests/conftest.py]: Why is this change needed?

Comment (RazTamir) [reply tests/conftest.py]: We want to delete the pool only if that's cephblockpool.
cephfilesystem shouldn't be removed

File: tests/conftest.py (hunk 1 above)
Comment (RazTamir) [reply tests/conftest.py]: I'm not sure it ever worked like that cause if you providing custom_data, you expect that all information would be there and therefore, you don't have storageclass initialized or project.

Is that correct?

**Comment by RazTamir** [comment]
Verified:

tests/e2e/test_pvc_creation_performance.py::TestPVCCreationPerformance::test_pvc_creation_measurement_performance[CephFS] 
-------------------------------------------------------------------------------------- live log setup ---------------------------------------------------------------------------------------
18:47:47 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig get Pod --selector=app=rook-ceph-tools -o yaml
18:47:48 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig rsh rook-ceph-tools-5f5dc75fd5-ht2k9 ceph auth get-key client.admin --format json-pretty
18:47:51 - MainThread - ocs_ci.ocs.resources.ocs - INFO - Adding Secret with name secret-test-cephfs-1418475146
18:47:51 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig create -f /tmp/Secretzp3cgfzt -o yaml
18:47:52 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig get CephFileSystem ocsci-cephfs -o yaml
18:47:54 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig get CephFileSystem  -o yaml
18:47:55 - MainThread - ocs_ci.ocs.resources.ocs - INFO - Adding StorageClass with name storageclass-test-cephfs-1418475577
18:47:55 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig create -f /tmp/StorageClassbwf4jj67 -o yaml
--------------------------------------------------------------------------------------- live log call ---------------------------------------------------------------------------------------
18:47:56 - MainThread - tests.e2e.test_pvc_creation_performance - INFO - Start creating new PVC
18:47:56 - MainThread - ocs_ci.ocs.resources.ocs - INFO - Adding PersistentVolumeClaim with name pvc-test-1418475688
18:47:56 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig create -f /tmp/PersistentVolumeClaim73zgb1mu -o yaml
18:47:58 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig get Pod --selector=app=csi-rbdplugin-provisioner -o yaml
18:47:59 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig logs csi-rbdplugin-provisioner-0 -c csi-provisioner
FAILED
------------------------------------------------------------------------------------- live log teardown -------------------------------------------------------------------------------------
18:48:03 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig delete PersistentVolumeClaim pvc-test-1418475688
18:48:07 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig get PersistentVolumeClaim pvc-test-1418475688 -o yaml
18:48:08 - MainThread - ocs_ci.ocs.ocp - INFO - PersistentVolumeClaim pvc-test-1418475688 got deleted successfully
18:48:08 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig delete StorageClass storageclass-test-cephfs-1418475577
18:48:10 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig get StorageClass storageclass-test-cephfs-1418475577 -o yaml
18:48:12 - MainThread - ocs_ci.ocs.ocp - INFO - StorageClass storageclass-test-cephfs-1418475577 got deleted successfully
18:48:12 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig delete Secret secret-test-cephfs-1418475146
18:48:13 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/ratamir/ocs_cluster1/auth/kubeconfig get Secret secret-test-cephfs-1418475146 -o yaml


You can see here the creation of PVC in the test call and deletion in teardown call

**Comment by clacroix12** [comment]
This might be a silly question but why does this need to be a fixture? We're basically creating a `delete_resource()` function here, right?

**Comment by RazTamir** [comment]
> This might be a silly question but why does this need to be a fixture? We're basically creating a `delete_resource()` function here, right?

Right but ... :)
the creation of that resource (PVC in this case) is part of the test, and the deletion should be done in the teardown

**Comment by clacroix12** [comment]
> the creation of that resource (PVC in this case) is part of the test, and the deletion should be done in the teardown

Isn't that exactly what the factory fixture is for? What's stopping you from using the pvc_factory in this use case. You can create the pvc in the test using the factory and the finalizer will handle cleanup of the resource in 'teardown.' Not sure what use-case this teardown factory is trying to cover.

**Comment by RazTamir** [comment]
The factory fixtures we currently have are not covering all resources we can create (daemon sets, deployments, monitoring resources, etc.) and that teardown fixture comes to help us.
In addition, it can provide great flexibility to keep using the helpers.py functions and still get the benefit of having the teardown clean up everything in the teardown

**Comment by clacroix12** [comment]
@RazTamir That makes sense, thanks! I think if we have a certain resource that we're doing this with a lot it's probably a good sign that the resource needs a fixture of it's own, but allowing resources we don't have fixtures for yet benefit from being cleaned up in the teardown is nice.

File: tests/conftest.py [hunk 3]
```diff
@@ -354,6 +356,36 @@ def finalizer():
     return factory
 
 
+@pytest.fixture()
+def teardown_factory(request):
+    """
+    Tearing down a resource that was created during the test
```
Comment (vasukulkarni) [inline tests/conftest.py]: Can we add more docstring here on where/how to use this? I am not clear on the usecase

File: tests/conftest.py [hunk 4]
```diff
@@ -354,6 +356,36 @@ def finalizer():
...
+        """
+        Args:
+            resource_obj (OCS object): Object to teardown after the test
+
+        """
+        instances.append(resource_obj)
+
+    def finalizer():
+        """
+        Delete the Pod
```
Comment (vasukulkarni) [inline tests/conftest.py]: If I understand you are talking about generic cleanup of uncleaned objects? not just Pod right?

File: tests/conftest.py (hunk 2 above)
Comment (vasukulkarni) [reply tests/conftest.py]: @RazTamir this should be a different PR in itself, its like add on additional fix along with teardown factory PR :)

File: tests/conftest.py (hunk 3 above)
Comment (RazTamir) [reply tests/conftest.py]: done.

Please let me know if I need to add more

File: tests/conftest.py (hunk 4 above)
Comment (RazTamir) [reply tests/conftest.py]: Thanks!

File: tests/conftest.py (hunk 1 above)
Comment (fbalak) [reply tests/conftest.py]: You can have reference to name of some object in custom_data. You may want to keep the reference to object with that name also in object using it. However, it's not mandatory. If you don't provide parameter with the reference then the reference to object is empty.

If you want to change this functionality then I propose to do it in different PR as this PR is about a new fixture. Also it would need to be changed like this in all fixtures.

File: tests/conftest.py (hunk 2 above)
Comment (RazTamir) [reply tests/conftest.py]: fixed in https://github.com/red-hat-storage/ocs-ci/pull/572

File: tests/conftest.py (hunk 1 above)
Comment (RazTamir) [reply tests/conftest.py]: ok reverting this change and let's discuss this in new PR

File: tests/e2e/test_pvc_creation_performance.py [hunk 5]
```diff
@@ -18,26 +18,30 @@ class TestPVCCreationPerformance(E2ETest):
     pvc_size = '1Gi'
 
     @pytest.fixture()
-    def base_setup(self, request, storageclass_factory):
+    def base_setup(
+        self, request, interface_iterate, storageclass_factory
```
[the remark this answers is not in the document]
Comment (RazTamir) [reply tests/e2e/test_pvc_creation_performance.py]: fixed

File: tests/conftest.py [hunk 6]
```diff
@@ -354,6 +354,43 @@ def finalizer():
     return factory
 
 
+@pytest.fixture()
+def teardown_factory(request):
+    """
+    Tearing down a resource that was created during the test
+    To use this factory, you'll need to pass 'teardown_factory' to you test
```
Comment (vasukulkarni) [inline tests/conftest.py]: thanks for the docstring, one minor - s/you/your and one more instance below.  When you push new changes, I will use previous approval and merge it.

Comment (RazTamir) [reply tests/conftest.py]: done
